File: Topological_Sorting/program_order.cpp
#include <bits/stdc++.h>
using namespace std;

const int MX = 1002;
vector<int> adj[MX];
int indeg[MX];
int n, m;

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> m;  
    int no, u, v;
    for (int i = 0; i < m; i++){
        cin >> no;
        if (no == 0) continue;
        cin >> u;
        while(--no){
            cin >> v;
            adj[u].push_back(v);
            indeg[v]++;
            swap(u, v);
        }
    }

    // Topological sorting
    queue<int> Q;
    vector<int> res;
    for (int i=1; i<=n; i++){
        if (!indeg[i]) Q.push(i);
    }
    while(!Q.empty()){
        int cur = Q.front(); Q.pop();
        res.push_back(cur);

        for (int nxt : adj[cur]){
            indeg[nxt]--;
            if (indeg[nxt]==0) Q.push(nxt);
        }
    }
    if (res.size() != n){
        cout << 0;
        }
    else{
            for (auto s : res) 
                cout << s << " ";
        }
}

File: Topological_Sorting/program_order.py
from collections import deque 

MX = 10
adj = [[] for _ in range(MX)]
indeg = [0]*MX

def topological():
    Q = deque()
    res = []
    
    for i in range(1, n+1):
        if (indeg[i] == 0):
            Q.append(i)
    
    while (Q):
        cur = Q.popleft()
        res.append(cur)
        
        for nxt in adj[cur]:
            indeg[nxt] -= 1
            if (indeg[nxt] == 0):
                Q.append(nxt)
    
    if (len(res) != n):
        print("Cycle existe!")
    else:
        for e in res :
            print(e, end = " ")

n = 6   # num_node
order_list = [[1,4,3],[6,2,5,4],[2,3]]

for m in order_list:
    for i in range(1, len(m)):
        a, b = m[i-1:i+1]   # a -> b
        adj[a].append(b)
        indeg[b] += 1

topological()
        
